### PR TITLE
add v2.2 manifest list template

### PR DESCRIPTION
**Pull request by loosebazooka**
part of #1567

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  private List<ManifestDescriptorTemplate> manifests = new ArrayList<>();
+
+  @VisibleForTesting
+  List<ManifestDescriptorTemplate> getManifests() {
+    return manifests;
+  }
+
+  public List<String> getDigestForPlatform(String architecture, String os) {
```
Comment (TadCordle) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: `getDigestsForPlatform`?

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    return manifests
+        .stream()
+        .filter(
+            manifest -> {
```
Comment (TadCordle) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: would it look better to consolidate this to
```
manifest -> 
    manifest.platform != null 
        && os.equals(manifest.platform.os)
        && architecture.equals(manifest.platform.architecture
```
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  public static final String MANIFEST_MEDIA_TYPE =
+      "application/vnd.docker.distribution.manifest.list.v2+json";
+  private static final int SCHEMA_VERSION = 2;
+
+  @Override
+  public int getSchemaVersion() {
+    return SCHEMA_VERSION;
+  }
+
+  private List<ManifestDescriptorTemplate> manifests = new ArrayList<>();
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: Can this be `final`?

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+      @Nullable
+      @JsonProperty("os.version")
+      private String osVersion;
+
+      @Nullable
+      @JsonProperty("os.features")
+      private List<String> osFeatures;
+
+      @Nullable private String variant;
+      @Nullable private List<String> features;
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: Any rule for this type of field, given that you did `private List<ManifestDescriptorTemplate> manifests = new ArrayList<>()` above?

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+      }
+    }
+
+    @Nullable private String mediaType;
+    @Nullable private String digest;
+    private long size;
+    @Nullable private Platform platform;
+
+    /** Required by Jackson-JSON */
+    public ManifestDescriptorTemplate() {}
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: Interesting. Shouldn't this constructor be implicit and exist anyways?

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java (hunk 4 above)
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: I think features is technically nullable in that the json doesn't **need** to include it. However manifest **must** be defined in the json.

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java (hunk 5 above)
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: Yeah, for some reason I was getting an error with jackson, but that might be because I was using non-static inner classes. Let me try again.

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java (hunk 3 above)
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: I think it technically can, but it shouldn't be. Jackson is actually injecting some other list into there. The empty array list is a way to specify that this shouldn't be nullable?

Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: ended up re-organizing this.

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java (hunk 4 above)
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: I changed all this to be consistent.

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  public int getSchemaVersion() {
+    return SCHEMA_VERSION;
+  }
+
+  @Nullable private List<ManifestDescriptorTemplate> manifests;
+
+  @VisibleForTesting
+  List<ManifestDescriptorTemplate> getManifests() {
+    Preconditions.checkNotNull(manifests);
+    return manifests;
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: nit: `return Preconditions.checkNotNull(manifests)`??

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+   * @see <a href="https://github.com/opencontainers/image-spec/blob/master/descriptor.md">OCI
+   *     Content Descriptors</a>
+   */
+  static class ManifestDescriptorTemplate implements JsonTemplate {
+
+    static class Platform implements JsonTemplate {
+      @Nullable private String architecture;
+      @Nullable private String os;
+
+      // ignored properties
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: Are you going to use these optional fields soon in follow-ups? Otherwise, I think we can annotate the class with `@JsonIgnoreProperties(ignoreUnknown = true)` like in #1793?

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        .map(ManifestDescriptorTemplate::getDigest)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Template for inner JSON object representing content descriptor for a layer or container
+   * configuration.
+   *
+   * @see <a href="https://github.com/opencontainers/image-spec/blob/master/descriptor.md">OCI
+   *     Content Descriptors</a>
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: Sorry, I'm late to the party, but this link seems out of place since is a docker manifest list?

(If it is equally applicable to an OCI Image Index, maybe it should be broken out as a separate class?)

Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java]: Uh this javadoc needs to be cleaned up. Oops
